middleware/response.go

```diff
@@ -0,0 +1,157 @@
+	// The contract of 'Write' has it set headers and call WriteHeader under
+	// certain circmustances. We ignore that here, as we assume that all the
+	// underlying implementations do that for us.
+	m.mirrorHeaders()
+	for _, rw := range m.rws {
+		n, err := rw.Write(data)
+		if err != nil {
+			return n, err
+		}
+		if n < len(data) {
+			return n, io.ErrShortWrite
+		}
+	}
+	return len(data), nil
+}
+
+// WriteHeader writes the HTTP response header.
+func (m *MultiResponseWriter) WriteHeader(statusCode int) {
+	m.mirrorHeaders()
+	for _, rw := range m.rws {
+		rw.WriteHeader(statusCode)
+	}
+}
+
+// Hijack hijacks the first response writer that is a Hijacker.
+func (m *MultiResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	for _, rw := range m.rws {
+		hj, ok := rw.(http.Hijacker)
+		if ok {
+			return hj.Hijack()
+		}
+	}
+	return nil, nil, fmt.Errorf("MultiResponseWriter: can't cast any responses to Hijacker")
+}
+
+// badResponseLogger is an http.ResponseWriter that logs response headers and
+// some of the body when the response is erroneous (i.e. a 5xx).
+//
+// Using this means holding an extra copy of the request and response headers
+// in memory.
+type badResponseLogger struct {
+	recorder      *httptest.ResponseRecorder
+	statusCode    int
+	logBody       bool
+	bodyBytesLeft int
+}
+
+// newBadResponseLogger makes a new badResponseLogger.
+func newBadResponseLogger() *badResponseLogger {
+	return &badResponseLogger{
+		recorder:      httptest.NewRecorder(),
+		logBody:       false,
+		bodyBytesLeft: maxResponseBodyInLogs,
+		statusCode:    http.StatusOK,
+	}
+}
+
+func (b *badResponseLogger) dumpResponse() ([]byte, error) {
+	return httputil.DumpResponse(b.recorder.Result(), true)
+}
+
+// Header implements http.ResponseWriter.
+func (b *badResponseLogger) Header() http.Header {
+	return b.recorder.Header()
+}
+
+// WriteHeader implements http.ResponseWriter. It will immediately log the
+// response headers if `statusCode` is a 5XX.
+func (b *badResponseLogger) WriteHeader(statusCode int) {
+	b.statusCode = statusCode
+	b.recorder.WriteHeader(statusCode)
+	if 100 <= statusCode && statusCode < 500 {
+		return
+	}
+	b.logBody = true
+}
+
+// Write implements http.ResponseWriter. It will log the body up to
+// `maxResponseBodyInLogs` if the response is a 5XX.
+func (b *badResponseLogger) Write(data []byte) (int, error) {
+	// If we haven't written the headers yet, then Write is supposed to call
+	// WriteHeader with http.StatusOK. Since we don't want to do anything in
+	// that case (this struct is for *bad* responses only), we don't need to
+	// track whether or not we've called WriteHeader.
+	if !b.logBody {
+		// We don't need to log anything, so lie and say that we wrote
+		// everything, making this effectively a no-op.
+		return len(data), nil
+	}
+
+	if len(data) > b.bodyBytesLeft {
+		b.recorder.Write(data[:b.bodyBytesLeft])
+		b.recorder.WriteString("…")
```

Comment:
I don't think we want to do this more than once; maybe reset `logBody` at this point?